diff --git a/lisp/progmodes/executable.el b/lisp/progmodes/executable.el
index 00c898d261..6bad79f6e6 100644
--- a/lisp/progmodes/executable.el
+++ b/lisp/progmodes/executable.el
@@ -238,35 +238,36 @@ executable-set-magic
 
   (or buffer-read-only
       (if buffer-file-name
-	  (string-match executable-magicless-file-regexp
-			buffer-file-name))
+         (string-match executable-magicless-file-regexp
+                       buffer-file-name))
       (not (or insert-flag executable-insert))
       (> (point-min) 1)
       (save-excursion
-	(goto-char (point-min))
-	(add-hook 'after-save-hook 'executable-chmod nil t)
-	(if (looking-at "#![ \t]*\\(.*\\)$")
-	    (and (goto-char (match-beginning 1))
-		 ;; If the line ends in a space,
-		 ;; don't offer to change it.
-		 (not (= (char-after (1- (match-end 1))) ?\s))
-		 (not (string= argument
-			       (buffer-substring (point) (match-end 1))))
-		 (if (or (not executable-query) no-query-flag
-			 (save-window-excursion
-			   ;; Make buffer visible before question.
-			   (switch-to-buffer (current-buffer))
-			   (y-or-n-p (format-message
-				      "Replace magic number by `#!%s'? "
-				      argument))))
-		     (progn
-		       (replace-match argument t t nil 1)
-		       (message "Magic number changed to `#!%s'" argument))))
-	  (insert "#!" argument ?\n)
-	  (message "Magic number changed to `#!%s'" argument))))
-    interpreter)
-
-
+        (goto-char (point-min))
+        (add-hook 'after-save-hook 'executable-chmod nil t)
+        (if (looking-at "#![ \t]*\\(.*\\)$")
+            (let* ((beg (match-beginning 1))
+                   (end (match-end 1))
+                   (matched (buffer-substring beg end)))
+              (and (goto-char beg)
+                   ;; If the line ends in a space,
+                   ;; don't offer to change it.
+                   (not (= (char-after (1- end)) ?\s))
+                   (not (string= argument matched))
+                   (if (or (not executable-query) no-query-flag
+                           (save-window-excursion
+                             ;; Make buffer visible before question.
+                             (switch-to-buffer (current-buffer))
+                             (y-or-n-p (format-message
+                                        "Replace magic number by `#!%s'? "
+                                        argument))))
+                       (progn
+                         (delete-region beg end)
+                         (insert argument)
+                         (message "Magic number changed to `#!%s'" argument)))))
+          (insert "#!" argument ?\n)
+          (message "Magic number changed to `#!%s'" argument))))
+  interpreter)
 
 (defun executable-self-display ()
   "Turn a text file into a self-displaying Un*x command.
